subsys/usb/device_next/class/bt_hci.c

```diff
@@ -0,0 +1,491 @@
+
+	bi = udc_get_buf_info(buf);
+
+	if (bi->ep == bt_hci_get_bulk_out(c_nd)) {
+		return bt_hci_acl_out_cb(c_nd, buf, err);
+	}
+
+	if (bi->ep == bt_hci_get_bulk_in(c_nd) || bi->ep == bt_hci_get_int_in(c_nd)) {
+		k_sem_give(&hci_data->sync_sem);
+		return 0;
+	}
+
+	return usbd_ep_buf_free(uds_ctx, buf);
+}
+
+static void bt_hci_update(struct usbd_class_node *const c_nd,
+			  uint8_t iface, uint8_t alternate)
+{
+	LOG_DBG("New configuration, interface %u alternate %u",
+		iface, alternate);
+}
+
+static void bt_hci_enable(struct usbd_class_node *const c_nd)
+{
+	struct bt_hci_data *hci_data = c_nd->data->priv;
+
+	atomic_set_bit(&hci_data->state, BT_HCI_CLASS_ENABLED);
+	LOG_INF("Configuration enabled");
+	if (bt_hci_acl_out_start(c_nd)) {
+		LOG_ERR("Failed to start ACL OUT transfer");
+	}
+}
+
+static void bt_hci_disable(struct usbd_class_node *const c_nd)
+{
+	struct bt_hci_data *hci_data = c_nd->data->priv;
+
+	atomic_clear_bit(&hci_data->state, BT_HCI_CLASS_ENABLED);
+	LOG_INF("Configuration disabled");
+}
+
+static int bt_hci_ctd(struct usbd_class_node *const c_nd,
+		      const struct usb_setup_packet *const setup,
+		      const struct net_buf *const buf)
+{
+	struct net_buf *cmd_buf;
+
+	/* We expect host-to-device class request */
+	if (setup->RequestType.type != USB_REQTYPE_TYPE_CLASS) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	LOG_DBG("bmRequestType 0x%02x bRequest 0x%02x",
+		setup->bmRequestType, setup->bRequest);
+
+	cmd_buf = bt_buf_get_tx(BT_BUF_CMD, K_NO_WAIT, buf->data, buf->len);
+	if (!cmd_buf) {
+		LOG_ERR("Cannot get free buffer");
+		return -ENOMEM;
+	}
+
+	net_buf_put(&bt_hci_rx_queue, cmd_buf);
+
+	return 0;
+}
+
+static int bt_hci_init(struct usbd_class_node *const c_nd)
+{
+	return 0;
+}
+
+static struct usbd_class_api bt_hci_api = {
+	.request = bt_hci_request,
+	.update = bt_hci_update,
+	.enable = bt_hci_enable,
+	.disable = bt_hci_disable,
+	.control_to_dev = bt_hci_ctd,
+	.init = bt_hci_init,
+};
+
+#define BT_HCI_DESCRIPTOR_DEFINE(n)						\
+static struct usbd_bt_hci_desc bt_hci_desc_##n = {				\
+	.if0 = {								\
+		.bLength = sizeof(struct usb_if_descriptor),			\
+		.bDescriptorType = USB_DESC_INTERFACE,				\
+		.bInterfaceNumber = 0,						\
+		.bAlternateSetting = 0,						\
+		.bNumEndpoints = 3,						\
+		.bInterfaceClass = USB_BCC_WIRELESS_CONTROLLER,			\
+		.bInterfaceSubClass = USB_BT_HCI_SUBCLASS,			\
+		.bInterfaceProtocol = USB_BT_HCI_PROTOCOL,			\
+		.iInterface = 0,						\
+	},									\
+										\
+	.if0_int_ep = {								\
+		.bLength = sizeof(struct usb_ep_descriptor),			\
+		.bDescriptorType = USB_DESC_ENDPOINT,				\
+		.bEndpointAddress = 0x81,					\
+		.bmAttributes = USB_EP_TYPE_INTERRUPT,				\
+		.wMaxPacketSize = sys_cpu_to_le16(BT_HCI_DEFAULT_INT_EP_MPS),	\
+		.bInterval = 1,							\
+	},									\
+										\
+	.if0_out_ep = {								\
+		.bLength = sizeof(struct usb_ep_descriptor),			\
+		.bDescriptorType = USB_DESC_ENDPOINT,				\
+		.bEndpointAddress = 0x02,					\
```

Comment:
Not familiar with this: Why is this `bEndpointAddress` so low compared to the other 2?

Reply:
Because it is OUT, which means from device to host. The addresses 0x80-0x8F are IN, while 0x00-0x0F are OUT.

Universal Serial Bus Specification Revision 2.0, Table 9-13. Standard Endpoint Descriptor
> The address of the endpoint on the USB device described by this descriptor. The address is encoded as follows:
Bit 3...0: The endpoint number
Bit 6...4: Reserved, reset to zero
Bit 7: Direction, ignored for control endpoints
0 = OUT endpoint
1 = IN endpoint

Reply:
I see. 

This could be made clearer with defining the possible endpoint addresses and assigning one of them, instead of these "magic" numbers :)

Reply:
There is nothing magic, just basic USB knowledge.

Reply:
No offense, but that is a pretty poor argument :) Firstly the actual value is basically a struct like
```c
struct usb_ep_addr {
    uint8_t number:4,
    uint8_t reserved:3,
    uint8_t direction:1,
};
```

Secondly that argument also implies that the other defined values in this filed also don't need to be defined, as they are presumably also "just basic USB knowledge" :)

Reply:
No, until reserved is no longer reserved, what is very unlikely, there is no reason at all to define any macros. 0x81-0x8F | 0x01-0x0F are cross-system and concise in USB context. Also there are USB_CONTROL_EP_IN | USB_CONTROL_EP_OUT macros for 0x80 | 0x00 due to unique and specific nature of these endpoints.

Reply:
> Also there are USB_CONTROL_EP_IN | USB_CONTROL_EP_OUT macros for 0x80 | 0x00 due to unique and specific nature of these endpoints.

Shouldn't these be used then? I'd be happier to see `dir | number` (i.e. `USB_CONTROL_EP_OUT | 0x01`) than just knowing that BIT(7) is implicitly the direction :)

Reply:
Looking at all the other BT USB drivers, they actually define these endpoint addreses such as 
```c
#define BLUETOOTH_INT_EP_ADDR		0x81
#define BLUETOOTH_OUT_EP_ADDR		0x02
#define BLUETOOTH_IN_EP_ADDR		0x82
``` 
or
```c
#define BT_H4_OUT_EP_ADDR               0x01
#define BT_H4_IN_EP_ADDR                0x81
```

and also used in other non-BT drivers such as 
```c

#define RNDIS_INT_EP_ADDR		0x83
#define RNDIS_IN_EP_ADDR		0x82
#define RNDIS_OUT_EP_ADDR		0x01
```

So I still (strongly) request that we use defined values for these

Reply:
Multiple USB tools simply show the endpoints in 0x00-0x0F and 0x80-0x8F format. It is really commonplace in USB context to use the hex numbers directly.